x-pack/lib/monitoring/inputs/metrics/stats_event_factory.rb

```diff
@@ -7,14 +7,27 @@ class StatsEventFactory
     include ::LogStash::Util::Loggable
     require 'logstash/config/pipelines_info'
 
+    def self.format_cluster_uuid(stats)
+      result = stats.extract_metrics([:stats, :pipelines, :main, :config], :cluster_uuids)
+      if result && !result[:cluster_uuids].empty?
+        result[:cluster_uuids][0]
```

Comment:
I wonder about setting precedence here - we might want to have the `monitoring.cluster_uuid` override the value from `cluster_uuids`, particularly if there are >1 elasticsearch outputs pointing to different clusters.

We might want to log when we detect >1 distinct `cluster_uuid` values.

Reply:
I agree with you, explicit user defined values should have precedence on implicitly extracted values

Reply:
The Stack Monitoring UI is centered around the notion of an Elasticsearch cluster. User's first choose an Elasticsearch cluster, then drill deeper into Elasticsearch, Logstash, etc. monitoring data associated with that cluster. So the purpose of the `cluster_uuid` field in monitoring documents is to associate the monitoring data in the document with an ES cluster in the Stack Monitoring UI. 

In terms of Logstash pipelines with multiple ES outputs, I think it might make sense to show that Logstash pipeline in each of those ES clusters in the UI. The only way to pull this off today is to index the same document multiple times, but with a different cluster_uuid each time. We do this in the Metricbeat `logstash` module code: https://github.com/elastic/beats/blob/523528b7516adbda814d8770a2e259a4464023b0/metricbeat/module/logstash/node/data_xpack.go#L83-L91.

Happy to go a different direction here but either way we should probably keep the Metricbeat `logstash` module implementation and the implementation in this PR consistent.

Reply:
so we could use give precedence to the `monitoring.cluster_uuid` settings if defined us that. If not defined, collect the list of `cluster_uuid` from all ES output plugins and change the metric input to fire one event for each `cluster_uuid` discovered

Reply:
Actually, we are doing it the other way around in the Metricbeat `logstash` module. First precedence is given to cluster_uuid(s) discovered from the Logstash pipeline. If there are none, then second precedence is given to the `monitoring.cluster_uuid` setting in `logstash.yml`. If that is also empty, set the value to an empty string (`""`).

Reply:
@ycombinator Interesting! I agree with the first notion that Logstash data is shown for each of the clusters, good to keep consistency with the Logstash module.

For the second point, does this mean effectively that `monitoring.cluster_uuid` is only applicable when no elasticsearch output is defined? Do you log when `monitoring.cluster_uuid` is ignored?

Reply:
> For the second point, does this mean effectively that `monitoring.cluster_uuid` is only applicable when no elasticsearch output is defined?

Yes, that's how we've been treating it in the Metricbeat `logstash` module: `monitoring.cluster_uuid` is a fallback only in cases when something more specific (an output in a specific pipeline on the node) can't provide the cluster UUID.

> Do you log when `monitoring.cluster_uuid` is ignored?

We do not, but perhaps we could, at least at a `DEBUG` level. I'm happy to mirror in the Metricbeat `logstash` module whatever you decide to do about logging here in this PR.

Reply:
ok, cool. In that case, I think I would prefer that we log that (ideally only once, rather than each time a metric is sent...), and also document the fact that `monitoring.cluster_uuid` is only relevant when no elasticsearch output is defined, and will be ignored when at least one es out is present.

Reply:
I followed you suggestions. With commit eebfcd96709f18244c253515df2df525414c8f4e I switched the retrieval of `cluster_uuid` to be lazy and done only once in the input metric plugin instead of each event generation.

With commit 25beb497fcc468b674783512d5f9f88acbbfe1c3 I switched the cluster_uuid to be a list, retrieved by all elasticsearch output plugins, so that the same monitoring event is sent for each cluster_uuid.

@ycombinator I added a log line when we use the `cluster_uuids` instead of the setting from config file

What is left out is document better what's the intended use of `monitoring.cluster_uuid`setting.